### PR TITLE
Introduce and Implement `VestedTransfer` Trait

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR introduces a `VestedTransfer` Trait, which handles making a transfer while also applying a vesting schedule to that balance.

This can be used in pallets like the Treasury pallet, where now we can easily introduce a `vested_spend` extrinsic as an alternative to giving all funds up front.

We implement `()` for the `VestedTransfer` trait, which just returns an error, and allows anyone to opt out from needing to use or implement this trait.

This PR also updates the logic of `do_vested_transfer` to remove the "pre-check" which was needed before we had a default transactional layer in FRAME.

Finally, I also fixed up some bad formatting in the test.rs file.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
After this PR, I would like to open a PR to Treasury pallet to enable the option to do a `vested_spend`, and see how opengov uses it.

File: substrate/frame/vesting/src/lib.rs [hunk 1]
```diff
@@ -784,3 +776,24 @@ where
...
+	fn vested_transfer(
+		source: &T::AccountId,
+		target: &T::AccountId,
+		locked: BalanceOf<T>,
+		per_block: BalanceOf<T>,
+		starting_block: BlockNumberFor<T>,
+	) -> DispatchResult {
+		let schedule = VestingInfo::new(locked, per_block, starting_block);
+		let source = <T::Lookup as StaticLookup>::unlookup(source.clone());
+		let target = <T::Lookup as StaticLookup>::unlookup(target.clone());
```
Comment (bkchr) [inline substrate/frame/vesting/src/lib.rs:796]: Inside `do_vested_transfer` you then call `lookup`. TLDR, `do_vested_transfer` should directly take `T::AccountId`

File: substrate/frame/vesting/src/lib.rs [hunk 2]
```diff
@@ -534,32 +534,25 @@ impl<T: Config> Pallet<T> {
...
+		// If adding this vesting schedule fails, all storage changes are undone due to FRAME's
+		// default transactional layers.
+		Self::add_vesting_schedule(
 			&target,
 			schedule.locked(),
 			schedule.per_block(),
 			schedule.starting_block(),
-		);
-		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
+		)?;
```
Comment (bkchr) [inline substrate/frame/vesting/src/lib.rs:555]: This means that everyone calling this will need to use `transactional` to revert the changes if there is an error.

**Review by bkchr — APPROVED** [review]
Generally okay, but some things need to change.

File: substrate/frame/vesting/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply substrate/frame/vesting/src/lib.rs:555]: everyone in frame uses transactional? what am I missing?

All code inside the runtime now assume we can error at any point in the code. when this was written, it wasn't the case, hence the need to "double check"

File: substrate/frame/vesting/src/lib.rs (hunk 1 above)
Comment (shawntabrizi) [reply substrate/frame/vesting/src/lib.rs]: done

File: substrate/frame/support/src/traits/tokens/currency/lockable.rs [hunk 3]
```diff
@@ -112,3 +112,52 @@ pub trait VestingSchedule<AccountId> {
...
 	fn remove_vesting_schedule(who: &AccountId, schedule_index: u32) -> DispatchResult;
 }
+
+/// A vested transfer over a currency. This allows a transferred amount to vest over time.
+pub trait VestedTransfer<AccountId> {
+	/// The quantity used to denote time; usually just a `BlockNumber`.
+	type Moment;
+
+	/// The currency that this schedule applies to.
+	type Currency: Currency<AccountId>;
```
Comment (gui1117) [inline substrate/frame/support/src/traits/tokens/currency/lockable.rs:122]: I see a lot of PR moving from Currency to Fungible.
What are the reason to use one or the other? here we could use Fungible with Hold I guess.

File: substrate/frame/support/src/traits/tokens/currency/lockable.rs [hunk 4]
```diff
@@ -112,3 +112,52 @@ pub trait VestingSchedule<AccountId> {
...
+	//
+	// Example: Assume we are on block 100. If `locked` amount is 100, and `per_block` is 1:
+	// 	- If `starting_block` is 0, then the whole 100 tokens will be available right away as the
+	//    vesting schedule started in the past and has fully completed.
+	// 	- If `starting_block` is 50, then 50 tokens are made available right away, and 50 more
+	//    tokens will unlock one token at a time until block 150.
+	// 	- If `starting_block` is 100, then each block, 1 token will be unlocked until the whole
+	//    balance is unlocked at block 200.
+	// 	- If `starting_block` is 200, then the 100 token balance will be completely locked until
+	//    block 200, and then start to unlock one token at a time until block 300.
```
Comment (gui1117) [inline substrate/frame/support/src/traits/tokens/currency/lockable.rs:138]:
```suggestion
	/// Execute a vested transfer from `source` to `target` with the given schedule:
	/// 	- `locked`: The amount to be transferred and for the vesting schedule to apply to.
	/// 	- `per_block`: The amount to be unlocked each block. (linear vesting)
	/// 	- `starting_block`: The block where the vesting should start. This block can be in the past
	///    or future, and should adjust when the tokens become available to the user.
	///
	/// Example: Assume we are on block 100. If `locked` amount is 100, and `per_block` is 1:
	/// 	- If `starting_block` is 0, then the whole 100 tokens will be available right away as the
	///    vesting schedule started in the past and has fully completed.
	/// 	- If `starting_block` is 50, then 50 tokens are made available right away, and 50 more
	///    tokens will unlock one token at a time until block 150.
	/// 	- If `starting_block` is 100, then each block, 1 token will be unlocked until the whole
	///    balance is unlocked at block 200.
	/// 	- If `starting_block` is 200, then the 100 token balance will be completely locked until
	///    block 200, and then start to unlock one token at a time until block 300.
```

File: substrate/frame/vesting/src/tests.rs [hunk 5]
```diff
@@ -630,8 +633,10 @@ fn merge_ongoing_schedules() {
...
 			ED * 10,
-			ED,                          // Vest over 10 blocks.
-			sched0.starting_block() + 5, // Start at block 15.
+			// Vest over 10 blocks.
+			ED,
+			// Start at block 15.
+			sched0.starting_block() + 5,
 		);
 		assert_ok!(Vesting::vested_transfer(Some(4).into(), 2, sched1));
 		assert_eq!(VestingStorage::<Test>::get(&2).unwrap(), vec![sched0, sched1]);
```
Comment (gui1117) [inline substrate/frame/vesting/src/tests.rs:642]: it would be good to have one test with a call to `vested_transfer` implementation.

File: substrate/frame/vesting/src/lib.rs (hunk 2 above)
Comment (gui1117) [reply substrate/frame/vesting/src/lib.rs:549]: But `VestedTransfer::vested_transfer` function is not a pallet call. If someone implements an error recovery logic it can get messy I suppose.

Maybe it is standard practice now, but I would add a small comment on top of `do_vesting_transfer` that the storage is dirty in case the method becomes public in the future.
And in the doc of `VestedTransfer::vested_transfer` I would also write that the storage can be dirty when error is returned.

Comment (bkchr) [reply substrate/frame/vesting/src/lib.rs:549]: > But `VestedTransfer::vested_transfer` function is not a pallet call. If someone implements an error recovery logic it can get messy I suppose.

Exactly. If someone handles the error, it will not be transactional. Or you do the transactional handling explicitly in the implementation of the trait.

File: substrate/frame/support/src/traits/tokens/currency/lockable.rs (hunk 3 above)
Comment (shawntabrizi) [reply substrate/frame/support/src/traits/tokens/currency/lockable.rs:122]: Ideally we would change to fungible, but it would not be a trivial change due to the way that Freezes/Holds are managed versus the Currency Trait's Reserve/Locks.

It would really be a much bigger PR, and I don't think I have the energy for that one right now.

File: substrate/frame/vesting/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply substrate/frame/vesting/src/lib.rs:549]: Thanks, I really didn't consider this.

I feel there must be many traits in Polkadot SDK which are vulnerable to the same problem, right?

File: substrate/frame/vesting/src/tests.rs (hunk 5 above)
Comment (shawntabrizi) [reply substrate/frame/vesting/src/tests.rs:642]: done

File: substrate/frame/support/src/traits/tokens/currency/lockable.rs (hunk 3 above)
Comment (gui1117) [reply substrate/frame/support/src/traits/tokens/currency/lockable.rs:122]: Good to know, yes it should be done in another PR if any

File: substrate/frame/vesting/src/lib.rs [hunk 6]
```diff
@@ -525,41 +527,33 @@ impl<T: Config> Pallet<T> {
...
 
-		T::Currency::transfer(
-			&source,
-			&target,
-			schedule.locked(),
-			ExistenceRequirement::AllowDeath,
-		)?;
+		T::Currency::transfer(source, target, schedule.locked(), ExistenceRequirement::AllowDeath)?;
 
 		// We can't let this fail because the currency transfer has already happened.
```
Comment (gui1117) [inline substrate/frame/vesting/src/lib.rs:550]:
```suggestion
		// We can't let this fail because the currency transfer has already happened.
		// Must be successful as it has been checked before.
		// Better to return error on failure anyway.
```

File: substrate/frame/vesting/src/lib.rs (hunk 2 above)
Comment (gui1117) [reply substrate/frame/vesting/src/lib.rs:556]: I agree that in general people should probably rarely recover from a dispatch error, and it would be simpler to require the caller to rollback.

Maybe we can make a transition with a new type: `DispatchErrorDirty` or `DispatchErrorToRollback` which means that the storage is dirty and should be rollbacked to handle the error manually.

Or maybe everybody already assume that a `DispatchError` can have a dirty storage.

File: substrate/frame/vesting/src/lib.rs [hunk 7]
```diff
@@ -525,41 +527,33 @@ impl<T: Config> Pallet<T> {
...
-		let res = Self::add_vesting_schedule(
-			&target,
+		Self::add_vesting_schedule(
+			target,
 			schedule.locked(),
 			schedule.per_block(),
 			schedule.starting_block(),
-		);
-		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
```
Comment (gui1117) [inline substrate/frame/vesting/src/lib.rs:562]: this check wasn't so bad. you can do the assert in `map_err`, or do `res?;` after this line

**Review by gui1117 — APPROVED** [review]

File: substrate/frame/vesting/src/lib.rs [hunk 8]
```diff
@@ -525,40 +527,35 @@ impl<T: Config> Pallet<T> {
...
+		// Must be successful as it has been checked before.
+		// Better to return error on failure anyway.
 		let res = Self::add_vesting_schedule(
-			&target,
+			target,
 			schedule.locked(),
 			schedule.per_block(),
 			schedule.starting_block(),
-		);
+		)?;
```
Comment (shawntabrizi) [inline substrate/frame/vesting/src/lib.rs:558]:
```suggestion
		);
```

File: substrate/frame/vesting/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply substrate/frame/vesting/src/lib.rs]: yeah this kind of direction would be nice overall imo

**Review by bkchr — APPROVED** [review]
